src/services/TagKeysProviders.ts

```diff
@@ -0,0 +1,36 @@
+import { logger } from './logger';
+import { LokiQuery } from './lokiQuery';
+import { DataSourceGetTagKeysOptions, GetTagResponse, MetricFindValue } from '@grafana/data';
+import { AdHocFiltersVariable } from '@grafana/scenes';
+import { DataSourceWithBackend, getDataSourceSrv } from '@grafana/runtime';
+import { getDataSource } from './scenes';
+import { LABELS_TO_REMOVE } from './filters';
+import { joinTagFilters } from './query';
+
+export async function getLabelsTagKeysProvider(variable: AdHocFiltersVariable): Promise<{
+  replace?: boolean;
+  values: GetTagResponse | MetricFindValue[];
+}> {
+  const datasource_ = await getDataSourceSrv().get(getDataSource(variable));
+  if (!(datasource_ instanceof DataSourceWithBackend)) {
+    logger.error(new Error('getTagKeysProvider: Invalid datasource!'));
+    throw new Error('Invalid datasource!');
+  }
+  const datasource = datasource_ as DataSourceWithBackend<LokiQuery>;
+
+  if (datasource && datasource.getTagKeys) {
+    const filters = joinTagFilters(variable);
+
+    const options: DataSourceGetTagKeysOptions<LokiQuery> = {
+      filters,
+    };
+
+    const result = (await datasource.getTagKeys(options)) as MetricFindValue[];
```

Comment:
```suggestion
    const tagKeys = await datasource.getTagKeys(options);
    const result: MetricFindValue[] = Array.isArray(tagKeys) ? tagKeys : [];
```
This seems to do the trick. A bit less tech debt for when we're tasked to remove type assertions.